joeynmt/transformer.py

```diff
@@ -0,0 +1,233 @@
+        return self.linears[-1](x)
+
+
+class PositionwiseFeedForward(nn.Module):
+    """
+    Position-wise Feed-forward layer
+    Projects to d_ff and then back down to d_model.
+    """
+
+    def __init__(self, d_model, d_ff, dropout=0.1):
+        """
+        Initializes position-wise feed-forward layer.
+        :param d_model: dimensionality of the input.
+        :param d_ff: dimensionality of intermediate representation
+        :param dropout:
+        """
+        super(PositionwiseFeedForward, self).__init__()
+        self.layer = nn.Sequential(
+            nn.Linear(d_model, d_ff),
+            nn.ReLU(),
+            nn.Dropout(dropout),
+            nn.Linear(d_ff, d_model)
+        )
+
+    def forward(self, x):
+        return self.layer(x)
+
+
+class PositionalEncoding(nn.Module):
+    """
+    Pre-compute position encodings.
+    In forward pass, this adds the position-encodings to the
+    input for as many time steps as necessary.
+    """
+
+    def __init__(self, d_model, dropout, max_len=5000):
+        """
+
+        :param d_model: dimensionality of the model
+        :param dropout:
+        :param max_len: maximum length for which  to pre-compute PE
+        """
+        super(PositionalEncoding, self).__init__()
+        self.dropout = nn.Dropout(p=dropout)
+
+        # Compute the positional encodings once in log space.
+        pe = torch.zeros(max_len, d_model)
+        position = torch.arange(0, max_len, dtype=torch.float32).unsqueeze(1)
+        div_term = torch.exp(torch.arange(0, d_model, 2, dtype=torch.float32) *
+                             -(math.log(10000.0) / d_model))
+        pe[:, 0::2] = torch.sin(position * div_term)
+        pe[:, 1::2] = torch.cos(position * div_term)
+        pe = pe.unsqueeze(0)
+        self.register_buffer('pe', pe)
+
+    def forward(self, x):
+        """
+        Add PE to x.
+        :param x:
+        :return: x with position encodings added.
+        """
+        x = x + self.pe[:, :x.size(1)]
+        return self.dropout(x)
+
+
+class NoamScheduler:
+    """Optim wrapper that implements rate."""
+
+    def __init__(self, model_size, factor, warmup, optimizer):
+        self.optimizer = optimizer
+        self._step = 0
+        self.warmup = warmup
+        self.factor = factor
+        self.model_size = model_size
+        self._rate = 0
+
+    def step(self):
+        """Update parameters and rate"""
+        self._step += 1
+        rate = self.rate()
+        for p in self.optimizer.param_groups:
+            p['lr'] = rate
+        self._rate = rate
+
+    def rate(self, step=None):
+        """Implement `lrate` above"""
+        if step is None:
+            step = self._step
+        return self.factor * \
+               (self.model_size ** (-0.5) *
+                min(step ** (-0.5), step * self.warmup ** (-1.5)))
+
+    #pylint: disable=no-self-use
+    def state_dict(self):
+        return None
+
+
+class LabelSmoothing(nn.Module):
```

Comment:
Does it work for RNNs as well?

Reply:
It should but we're not using it yet anywhere right now.

Reply:
Working on it.

Reply:
It does now.